internal/resolver/dns/dns_resolver.go

```diff
@@ -201,18 +207,49 @@ func (d *dnsResolver) Close() {
 
 func (d *dnsResolver) watcher() {
 	defer d.wg.Done()
+	backoffIndex := 1
+	nextBackoffRate := backoff.DefaultExponential.Backoff(backoffIndex)
 	for {
 		select {
 		case <-d.ctx.Done():
+			d.pollTimer.Stop()
 			return
 		case <-d.rn:
+			// Reset polling timer if active, as ResolveNow was called from ClientConn.
+			if !d.pollTimer.Stop() {
+				select {
+				case <-d.pollTimer.C:
+				default:
+				}
+			}
+			backoffIndex = 1
+			nextBackoffRate = backoff.DefaultExponential.Backoff(backoffIndex)
+		case <-d.pollTimer.C:
 		}
 
 		state, err := d.lookup()
 		if err != nil {
+			// Report error to the underlying grpc.ClientConn.
 			d.cc.ReportError(err)
 		} else {
-			d.cc.UpdateState(*state)
+			err = d.cc.UpdateState(*state)
+		}
+
+		if err == nil {
+			// Success; reset polling for next usage.
+			if !d.pollTimer.Stop() {
+				select {
+				case <-d.pollTimer.C:
+				default:
+				}
+			}
+			backoffIndex = 1
+			nextBackoffRate = backoff.DefaultExponential.Backoff(backoffIndex)
+		} else {
+			// Start polling on an error found in DNS Resolver or an error received from ClientConn.
+			d.pollTimer = newTimer(nextBackoffRate)
```

Comment:
I'm not sure if there is a big benefit here, but you can `Reset` the timer to reuse it.  Note that if you aren't going to `Reset` and reuse, then don't bother with the draining of the timer channel.  That's only useful if you are reusing it.

Reply:
Thought about this. Reset I think is useful on the success case Resetting it to max Duration. On the failure case, I think it is best to construct new memory for two reasons: one: Able to test it with global variable NewTimer. two: if the polling timer is active and the resolver receives another ResolveNow{} call, I would prefer it instead of draining and calling Reset on the failure case, to construct a new one instead and move forward to next iteration.